tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -6839,7 +6839,12 @@ const Expr *SpirvEmitter::collectArrayStructIndices(
     // The index into an array must be an integer number.
     const auto *idxExpr = indexing->getIdx();
     const auto idxExprType = idxExpr->getType();
-    SpirvInstruction *thisIndex = doExpr(idxExpr);
+    if (idxExpr->isLValue()) {
+      emitError("Index of ArraySubscriptExpr must be rvalue",
+                idxExpr->getExprLoc());
+      return nullptr;
+    }
+    SpirvInstruction *thisIndex = loadIfGLValue(idxExpr);
```

Comment:
I'm a bit confused here. If `idxExpr` is LValue, then the code errors out above, right? so does this `loadIfGLValue` get executed if `idxExpr` is an LValue?

Reply:
Good question. I believe `idxExpr` should never be LValue. If it is LValue, it means the given HLSL code or the AST parser has a bug, because we do not use LValue for the index of an array.

The reason why we use `loadIfGLValue` instead of `doExpr` is because we have some cases that `doExpr` returns SPIR-V instruction with LValue for `idxExpr` with RValue e.g., missing `ImplicitCast<LValueToRValue>` #3620 

I added comments to explain it.